### PR TITLE
MES-2617 autosave result

### PR DESCRIPTION
**Pull request by edwardwilson**
## Description
Test submission effect now also looks for test data in the write-up state. Those in the WriteUp state will be sent as a partial save. Once saved the state will be moved to Autosaved.

This PR includes all the needed changes to the effects and state for autosave.

## Checklist

- [x] PR title includes the JIRA ticket number
- [ ] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [x] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

## Screenshots (optional)

### CODE REVIEW

**Comment by roberthall12** [comment]
Can you add a quick description for what's been added?

File: src/providers/test-submission/__tests__/test-submission.spec.ts [hunk 1]
```diff
@@ -57,12 +58,13 @@ describe('TestSubmissionProvider', () => {
         index: 0,
         slotId: '',
         payload: {},
+        status: TestStatus.Completed,
       }] as TestToSubmit[]).subscribe();
 
-      httpMock.expectOne('https://www.example.com/api/v1/test-result');
+      httpMock.expectOne('https://www.example.com/api/v1/test-result?partial=false');
```
Comment (roberthall12) [inline src/providers/test-submission/__tests__/test-submission.spec.ts]: I don't think we need `partial=false` for a full submission, see https://wiki.dvsacloud.uk/display/MES/Autosave+LLD

File: src/providers/test-submission/test-submission.ts [hunk 2]
```diff
@@ -69,5 +81,8 @@ export class TestSubmissionProvider {
     };
     return removeNullFields(data);
   }
-
+  removeFieldsForPartialData = (data: StandardCarTestCATBSchema): Partial<StandardCarTestCATBSchema> => {
+    data.testSummary = null;
```
Comment (roberthall12) [inline src/providers/test-submission/test-submission.ts]: We can't do this as this will remove `D255`, `debriefWitnessed` from the state

**Review by roberthall12 — CHANGES_REQUESTED** [review]
Good - just need to modify the `removeFieldsForPartialData` function to allow us to keep the fields collected in pass/non-pass finalisation pages

File: src/modules/tests/tests.actions.ts [hunk 3]
```diff
@@ -78,6 +81,15 @@ export class SendCompletedTestsFailure implements Action {
   readonly type = SEND_COMPLETED_TESTS_FAILURE;
 }
 
+export class SendPartialTestSuccess implements Action {
+  readonly type = SEND_PARTIAL_TEST_SUCCESS;
+  constructor(public testId: string) {}
```
Comment (buzzub) [inline src/modules/tests/tests.actions.ts]: I think this should be named `slotId` instead. `testId` isn't really used anywhere else

File: src/modules/tests/tests.effects.ts [hunk 4]
```diff
@@ -252,15 +253,30 @@ export class TestsEffects {
...
+                ? new testActions.SendPartialTestsFailure()
+                : new testActions.SendCompletedTestsFailure();
             });
           }),
         );
     }),
   );
 
+  @Effect()
+  sendPartialTestsSuccessEffect$ = this.actions$.pipe(
```
Comment (buzzub) [inline src/modules/tests/tests.effects.ts]: The effect is plural but the action is singular so should this be `sendPartialTestSuccessEffect` instead?

**Comment by buzzub** [comment]
Will approve once Robert's comments have been addressed 👍

File: src/providers/test-submission/test-submission.ts [hunk 5]
```diff
@@ -37,10 +39,16 @@ export class TestSubmissionProvider {
...
+    const deepClonedData = cloneDeep(testToSubmit.payload);
+    const cleanData = this.removeNullFieldsDeep(
+      testToSubmit.status === TestStatus.WriteUp
+        ? this.removeFieldsForPartialData(deepClonedData)
+        : deepClonedData);
+
+    return this.httpClient.post(
+      this.buildUrl(testToSubmit.status),
+      this.compressData(cleanData),
       // Using cloneDeep() to prevent the initialState of the reducers from being modified
```
Comment (TomBillingtonUK) [inline src/providers/test-submission/test-submission.ts]: This comment is now in the wrong place

**Review by roberthall12 — APPROVED** [review]

**Review by buzzub — APPROVED** [review]

**Pull request merged**